geo/src/algorithm/concavehull.rs

```diff
@@ -0,0 +1,418 @@
+    where
+        T: Float + RTreeNum;
+}
+
+impl<T> ConcaveHull<T> for Polygon<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let points: Vec<_> = self.exterior().points_iter().collect();
+        Polygon::new(LineString::from(concave_hull(&points, concavity)), vec![])
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiPolygon<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let aggregated: Vec<Point<T>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.exterior().0.iter().map(|c| Point(*c)))
+            .collect();
+        Polygon::new(
+            LineString::from(concave_hull(&aggregated, concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for LineString<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        Polygon::new(
+            LineString::from(concave_hull(&self.clone().into_points(), concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiLineString<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let aggregated: Vec<Point<T>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.clone().into_points())
+            .collect();
+        Polygon::new(
+            LineString::from(concave_hull(&aggregated, concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiPoint<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        Polygon::new(
+            LineString::from(concave_hull(&self.0.clone(), concavity)),
+            vec![],
+        )
+    }
+}
+
+fn find_point_closest_to_line<T>(
+    interior_points_tree: &RTree<Point<T>>,
+    line: Line<T>,
+    max_dist: &T,
+) -> Option<Point<T>>
+where
+    T: Float + RTreeNum,
+{
+    let h = max_dist.add(max_dist.clone());
+    let w = line.euclidean_length() + h;
+    let two = T::add(T::one(), T::one());
+    let search_dist = T::div(T::sqrt(T::powi(w, 2) + T::powi(h, 2)), two);
+    let mut candidates = interior_points_tree
+        .locate_within_distance(line.centroid(), search_dist)
+        .peekable();
+    let peek = candidates.peek();
+    match peek {
+        None => None,
+        Some(&point) => Some(candidates.fold(*point, |acc, candidate| {
+            if line.euclidean_distance(&acc) > line.euclidean_distance(candidate) {
+                *candidate
+            } else {
+                acc
+            }
+        })),
+    }
+}
+
+// This takes significant inspiration from:
+// https://github.com/mapbox/concaveman/blob/54838e1/index.js#L11
+fn concave_hull<T>(points: &Vec<Point<T>>, concavity: T) -> Vec<Point<T>>
+where
+    T: Float + RTreeNum,
+{
+    let multipoint: MultiPoint<T> = points.clone().into();
```

Comment:
Similarly, is there a way to avoid cloning here? Once again, the error is below:
```
error[E0277]: the trait bound `geo_types::Point<T>: std::convert::From<&std::vec::Vec<geo_types::Point<T>>>` is not satisfied
   --> geo/src/algorithm/concavehull.rs:151:44
    |
151 |     let multipoint: MultiPoint<T> = points.into();
    |                                            ^^^^ the trait `std::convert::From<&std::vec::Vec<geo_types::Point<T>>>` is not implemented for `geo_types::Point<T>`
    |
    = help: the following implementations were found:
              <geo_types::Point<T> as std::convert::From<(T, T)>>
              <geo_types::Point<T> as std::convert::From<[T; 2]>>
              <geo_types::Point<T> as std::convert::From<geo_types::Coordinate<T>>>
    = note: required because of the requirements on the impl of `std::convert::Into<geo_types::Point<T>>` for `&std::vec::Vec<geo_types::Point<T>>`
    = note: required because of the requirements on the impl of `std::convert::From<&std::vec::Vec<geo_types::Point<T>>>` for `geo_types::MultiPoint<T>`
    = note: required because of the requirements on the impl of `std::convert::Into<geo_types::MultiPoint<T>>` for `&std::vec::Vec<geo_types::Point<T>>`
```

Reply:
I don't see an easy way to get rid of this one.